dotcom-rendering/src/model/enhanceTableOfContents.ts

```diff
@@ -0,0 +1,127 @@
+import { JSDOM } from 'jsdom';
+import type { TOCType } from 'src/types/frontend';
+
+const isHeading = (element: CAPIElement, level: 'H2' | 'H3' | 'H4') => {
+	if (
+		element._type !==
+			'model.dotcomrendering.pageElements.TextBlockElement' &&
+		element._type !==
+			'model.dotcomrendering.pageElements.SubheadingBlockElement'
+	)
+		return false;
+	const frag = JSDOM.fragment(element.html);
+	return frag.firstElementChild?.nodeName === level;
+};
+
+const isH2 = (
+	element: CAPIElement,
+): element is SubheadingBlockElement | TextBlockElement => {
+	return isHeading(element, 'H2');
+};
+
+const isH3 = (
+	element: CAPIElement,
+): element is SubheadingBlockElement | TextBlockElement => {
+	return isHeading(element, 'H3');
+};
+
+const isH4 = (
+	element: CAPIElement,
+): element is SubheadingBlockElement | TextBlockElement => {
+	return isHeading(element, 'H4');
+};
+
+const extractText = (
+	element: SubheadingBlockElement | TextBlockElement,
+): string => {
+	const frag = JSDOM.fragment(element.html);
+	if (!frag.firstElementChild) return '';
+	return frag.textContent?.trim() ?? '';
+};
+
+const hasH2s = (blocks: Block[]) => {
+	return blocks.find((block) => block.elements.find(isH2));
+};
+
+const hasH3s = (blocks: Block[]) => {
+	return blocks.find((block) => block.elements.find(isH3));
+};
+
+const hasH4s = (blocks: Block[]) => {
+	return blocks.find((block) => block.elements.find(isH4));
+};
+
+const hasInteractiveContentsElement = (blocks: Block[]): boolean => {
+	return !!blocks.find((block) =>
+		block.elements.find(
+			(element) =>
+				element._type ===
+				'model.dotcomrendering.pageElements.InteractiveContentsBlockElement',
+		),
+	);
+};
+
+export const enhanceTableOfContents = (
+	format: CAPIFormat,
+	blocks: Block[],
+): TOCType[] | undefined => {
+	if (
+		format.design !== 'AnalysisDesign' ||
```

Comment:
I wrote this but maybe the negation isn't super easy to follow here? I guess early returns are nice but...

Reply:
Early returns _are_ nice! I tend to find it more readable.